libraries/chain/controller.cpp

```diff
@@ -1246,14 +1246,63 @@ struct controller_impl {
+            // if chain_head is legacy, update to non-legacy chain_head, this is needed so that the correct block_state is created in apply_block
+            if (std::holds_alternative<block_state_legacy_ptr>(chain_head.internal())) {
+               auto prev = forkdb.get_block(bsp->previous(), include_root_t::yes);
+               assert(prev);
+               chain_head = block_handle{prev};
+            }
+            apply_block(br, bsp, controller::block_status::complete, trx_meta_cache_lookup{});
+         } else {
+            // only called during transition when not a proper savanna block
+            fork_db.apply_l<void>([&](const auto& forkdb_l) {
+               block_state_legacy_ptr legacy = forkdb_l.get_block(bsp->id());
+               fork_db.switch_to_legacy(); // apply block uses to know what types to create
+               fc::scoped_exit<std::function<void()>> e([&]{fork_db.switch_to_both();});
+               apply_block(br, legacy, controller::block_status::complete, trx_meta_cache_lookup{});
+               // irreversible apply was just done, calculate new_valid here instead of in transition_to_savanna()
+               assert(legacy->action_receipt_digests);
+               auto prev = forkdb.get_block(legacy->previous(), include_root_t::yes);
+               assert(prev);
+               transition_add_to_savanna_fork_db(forkdb, legacy, bsp, prev);
+            });
+         }
+      }
+   }
+
+   void transition_add_to_savanna_fork_db(fork_database_if_t& forkdb,
+                                          const block_state_legacy_ptr& legacy, const block_state_ptr& new_bsp,
+                                          const block_state_ptr& prev) {
+      // legacy_branch is from head, all will be validated unless irreversible_mode(),
+      // IRREVERSIBLE applies (validates) blocks when irreversible, new_valid will be done after apply in log_irreversible
+      assert(read_mode == db_read_mode::IRREVERSIBLE || legacy->action_receipt_digests);
+      if (legacy->action_receipt_digests) {
+         auto action_mroot = calculate_merkle(*legacy->action_receipt_digests);
+         // Create the valid structure for producing
+         new_bsp->valid = prev->new_valid(*new_bsp, action_mroot);
+      }
+      forkdb.add(new_bsp, legacy->is_valid() ? mark_valid_t::yes : mark_valid_t::no, ignore_duplicate_t::yes);
+   }
+
    void transition_to_savanna() {
       assert(chain_head.header().contains_header_extension(instant_finality_extension::extension_id()));
       // copy head branch branch from legacy forkdb legacy to savanna forkdb
       fork_database_legacy_t::branch_t legacy_branch;
       block_state_legacy_ptr legacy_root;
       fork_db.apply_l<void>([&](const auto& forkdb) {
          legacy_root = forkdb.root();
-         legacy_branch = forkdb.fetch_branch(forkdb.head()->id());
+         legacy_branch = forkdb.fetch_branch(fork_db_head(forkdb, irreversible_mode())->id());
```

Comment:
Every time we call `fork_db_head` we pass `irreversible_mode())`, so why not just call `irreversible_mode())` inside `fork_db_head` instead of passing an extra parameter.

Reply:
I think it makes it more obvious. If `fork_db_head()` sometimes returns `fork_db` `pending_head()`, seems like that could really confuse someone. I would expect `fork_db_head()` to be simply `return fork_db->head()`.

Reply:
I'm not sure that it is beneficial to add redundency to be more explicit. 
Maybe we could use another function name `fork_db_head_or_pending_head()`?